### PR TITLE
allow reading dimensions from md groups

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**
- [X] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [X] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

Added the wrapper for the dimension call for md groups.

### CODE REVIEW

File: src/raster/mdarray.rs [hunk 1]
```diff
@@ -413,7 +424,7 @@ impl<'a> Group<'a> {
         }
     }
 
-    pub fn open_group(&self, name: &str, options: CslStringList) -> Result<Group> {
+    pub fn open_group(&'_ self, name: &str, options: CslStringList) -> Result<Group<'a>> {
```
Comment (lnicola) [inline src/raster/mdarray.rs:427]: I'm a bit groggy right now, does `self` here have a different lifetime from the resulting `Group`?

File: src/raster/mdarray.rs [hunk 2]
```diff
@@ -440,13 +451,45 @@ impl<'a> Group<'a> {
             Ok(Attribute::from_c_attribute(c_attribute))
         }
     }
+
+    pub fn dimensions(&self, options: CslStringList) -> Result<Vec<Dimension>> {
+        unsafe {
```
Comment (lnicola) [inline src/raster/mdarray.rs:456]: Please try to make this `unsafe` block smaller :-).

File: src/raster/mdarray.rs [hunk 3]
```diff
@@ -440,13 +451,45 @@ impl<'a> Group<'a> {
...
+            let mut num_dimensions: usize = 0;
+            let c_dimensions = GDALGroupGetDimensions(
+                self.c_group,
+                std::ptr::addr_of_mut!(num_dimensions),
```
Comment (lnicola) [inline src/raster/mdarray.rs:460]: I'm not necessarily against it, but I'd expect to see `addr_of_mut!` used with uninitialized variables, so it's a bit confusing here.

File: src/raster/mdarray.rs [hunk 4]
```diff
@@ -90,16 +96,21 @@ impl<'a> MDArray<'a> {
             let c_dimensions =
                 GDALMDArrayGetDimensions(self.c_mdarray, std::ptr::addr_of_mut!(num_dimensions));
 
-            if c_dimensions.is_null() {
-                return Err(_last_null_pointer_err("GDALMDArrayGetDimensions"));
+            // `num_dimensions` is `0`, we can safely return an empty vector
+            // `GDALMDArrayGetDimensions` does not state that errors can occur
+            if num_dimensions > 0 && c_dimensions.is_null() {
+                return Err(_last_null_pointer_err("GDALGroupGetDimensions"));
```
Comment (lnicola) [inline src/raster/mdarray.rs:102]: Maybe

```suggestion
                return Err(_last_null_pointer_err("GDALMDArrayGetDimensions"));
```

?

File: src/raster/mdarray.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply src/raster/mdarray.rs:427]: Yes, I prevented Group from having the lifetime of self. It needs the lifetime of the dataset.
I tested that in the changed test where I dropped a super-group.

File: src/raster/mdarray.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply src/raster/mdarray.rs]: [952eb06](https://github.com/georust/gdal/pull/291/commits/952eb0639407f5ed5402561d79268d82c1b3a4f3)

File: src/raster/mdarray.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply src/raster/mdarray.rs:455]: [c482412](https://github.com/georust/gdal/pull/291/commits/c4824127bcb594f53e4c23b2de03b32504fe1d7a)

File: src/raster/mdarray.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply src/raster/mdarray.rs]: [ccfa51d](https://github.com/georust/gdal/pull/291/commits/ccfa51d913579b45473c530b051b1249bf807a32)

**Comment by jdroenner** [comment]
can you resolve the conflicts?

**Comment by jdroenner** [comment]
i think this PR is good to merge. @lnicola would that be okay for you? Or are there open issues?

**Comment by lnicola** [comment]
bors r=jdroenner,lnicola

**Pull request merged**